src/storage/src/render/upsert.rs

```diff
@@ -211,22 +227,42 @@ where
                         a_ts == b_ts && a_key == b_key
                     });
 
+                    // Upsert the values into `commands_state`, by recording the latest
+                    // value (or deletion). These will be synced at the end to the `state`.
                     while let Some((ts, key, _, value)) = commands.next() {
+                        let command_state = commands_state
+                            .get_mut(&key)
+                            .expect("key missing from commands_state");
                         match value {
                             Some(value) => {
-                                if let Some(old_value) = state.insert(key, value.clone()) {
+                                if let Some(old_value) = command_state.take() {
```

Comment:
nit: `command_state.replace(value.clone())` exists in `Option` and does the same as the `insert` on the hashmap!